### PR TITLE
Provide more context in error messages

### LINKED ISSUE #7468: cabal-install-solver: improving error messages for misspelled language extensions
After helping someone debug a cabal build issue which was caused by a misspelled language extension in their `project.cabal` file I thought of an idea for improving the error message for misspelled language extensions.

Currently the error is `rejecting: test-cabal-0.1.0.0 (conflict: requires FlexibleConstraints)` where `FlexibleConstraints` is the misspelled language extension. I was thinking if we calculated the [Levenshtein distance](https://en.wikipedia.org/wiki/Levenshtein_distance) of the misspelled language extension and compared it to all known extensions https://github.com/haskell/cabal/blob/f09b8556e08bbaa3aa5acfb4d68e01c49e13d79a/Cabal/src/Language/Haskell/Extension.hs#L123 we could give more useful feedback. For example the error message could instead be something like
`rejecting: test-cabal-0.1.0.0 (conflict: requires FlexibleConstraints did you mean FlexibleContexts?)`From peeking a little at the code: https://github.com/haskell/cabal/blob/6896c6aa0e4804913aaba0bbbe00649e18f17bb8/cabal-install-solver/src/Distribution/Solver/Modular/Message.hs#L218 this seems trivial to implement, but maybe someone more experienced with the cabal codebase could comment on my idea before I submit a PR for this.

### PR DESCRIPTION
**Pull request by skykanin**
Error messages for misspelled extensions and languages now provide additional context and a suggested fix. Fixes  #7468

Examples:
```
rejecting: test-cabal-0.1.0.0 (conflict: requires unknown language
Haskell2015 did you mean Haskell2010?)
```
```
rejecting: test-cabal-0.1.0.0 (conflict: requires unknown extension
MultiParamClasses did you mean MultiParamTypeClasses?)
```
However it will not suggest a language or extension if the misspelled ones are too different from any of the valid languages or extensions

```
rejecting: test-cabal-0.1.0.0 (conflict: requires unknown extension AbsolutelyNotAnExtension)
```
```
rejecting: test-cabal-0.1.0.0 (conflict: requires unknown language AbsolutelyNotALanguage)
```
These changes have been tested through a mock cabal project and included unit tests.

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [x] The documentation has been updated, if necessary.

### CODE REVIEW

**Comment by phadej** [comment]
Test with `NoMonoLoclBinds` i.e. disabled extensions (which are rare but not unseen).

**Comment by skykanin** [comment]
> Test with `NoMonoLoclBinds` i.e. disabled extensions (which are rare but not unseen in cabal files too).

The error you get is:
```
rejecting: test-cabal-0.1.0.0 (conflict: requires unknown extension
NoMonoLoclBinds did you mean MonoLocalBinds?)
```

**Comment by Mikolaj** [comment]
> The error you get is:

Is that correct? Isn't `NoMonoLocalBinds` closer to `NoMonoLoclBinds` than `MonoLocalBinds`? And isn't it more likely to be what the user intended?

**Comment by skykanin** [comment]
> > The error you get is:
> 
> Is that correct? Isn't `NoMonoLocalBinds` closer to `NoMonoLoclBinds` than `MonoLocalBinds`? And isn't it more likely to be what the user intended?

ah right, I think this is because `NoMonoLocalBinds` isn't a part of the `KnownExtensions`type which is where I'm looking for the extension names. I'll have to iron that out.

**Comment by skykanin** [comment]
> > Test with `NoMonoLoclBinds` i.e. disabled extensions (which are rare but not unseen in cabal files too).
> 
> The error you get is:
> 
> ```
> rejecting: test-cabal-0.1.0.0 (conflict: requires unknown extension
> NoMonoLoclBinds did you mean MonoLocalBinds?)
> ```
> 
> I'll add this as a test case in my unit tests.

This is now fixed and you get the correct suggestion for disabled extensions
```
rejecting: test-cabal-0.1.0.0 (conflict: requires unknown extension
NoMonoLoclBinds did you mean NoMonoLocalBinds?)
```
New test cases for disabled extensions also included.

File: cabal-install-solver/src/Distribution/Solver/Modular/MessageUtils.hs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+showUnsupportedLanguage :: Language -> String
+showUnsupportedLanguage (UnknownLanguage langStr) = formatMessage cutoffRange "language" langStr (mostSimilarElement langStr (show <$> knownLanguages))
+showUnsupportedLanguage knownLanguage = prettyShow knownLanguage ++ "which is not supported"
+
+allKnownExtensions :: [String]
+allKnownExtensions = enabledExtensions ++ disabledExtensions
+  where
+    enabledExtensions = show <$> knownExtensions
+    disabledExtensions = ("No" ++) . show <$> knownExtensions
```
Comment (phadej) [inline cabal-install-solver/src/Distribution/Solver/Modular/MessageUtils.hs:33]: use `prettyShow` on `Extension`, i.e. `map (prettyShow . EnableExtension) knownExtensions` and `DisableExtension`. Using `"No"++` is wrong.

File: cabal-install-solver/src/Distribution/Solver/Modular/MessageUtils.hs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    mapDist element = (element, distance unknownElement element)
+
+-- Cutoff range for giving a suggested spelling
+cutoffRange :: Int
+cutoffRange = 10
+
+formatMessage :: Int -> String -> String -> String -> String
+formatMessage range elementType element suggestion
+  | withinRange range element suggestion = printf "unknown %s %s did you mean %s?" elementType element suggestion
+  | otherwise = printf "unknown %s %s" elementType element
```
Comment (phadej) [inline cabal-install-solver/src/Distribution/Solver/Modular/MessageUtils.hs:53]: `printf` isn't really used in `cabal-install` to concate the messages. Use `concat [...]` or `unwords`. There are examples like

```haskell
            die' verbosity $ unwords
              [ "Failed to download", show uri
              , ": SHA256 don't match; expected:", BS8.unpack (Base16.encode expected)
              , "actual:", BS8.unpack (Base16.encode actual)
              ]

```

in the code already.

File: cabal-install-solver/src/Distribution/Solver/Modular/MessageUtils.hs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  where
+    mapDist element = (element, distance unknownElement element)
+
+-- Cutoff range for giving a suggested spelling
+cutoffRange :: Int
+cutoffRange = 10
+
+formatMessage :: Int -> String -> String -> String -> String
+formatMessage range elementType element suggestion
+  | withinRange range element suggestion = printf "unknown %s %s did you mean %s?" elementType element suggestion
```
Comment (phadej) [inline cabal-install-solver/src/Distribution/Solver/Modular/MessageUtils.hs:52]: I'd put a period or `;` after an element word.

> (Unknown extension NoMonoLoclBInds. Did you mean NoMonoLocalBinds?)

Comment (skykanin) [reply cabal-install-solver/src/Distribution/Solver/Modular/MessageUtils.hs]: Fixed

File: cabal-install-solver/src/Distribution/Solver/Modular/MessageUtils.hs (hunk 2 above)
Comment (skykanin) [reply cabal-install-solver/src/Distribution/Solver/Modular/MessageUtils.hs]: Fixed

File: cabal-install-solver/src/Distribution/Solver/Modular/MessageUtils.hs (hunk 1 above)
Comment (skykanin) [reply cabal-install-solver/src/Distribution/Solver/Modular/MessageUtils.hs]: Fixed

**Review by phadej — APPROVED** [review]

File: cabal-install-solver/src/Distribution/Solver/Modular/MessageUtils.hs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+import Data.Foldable (minimumBy)
+import Data.Ord (comparing)
+import Distribution.Pretty (prettyShow) -- from Cabal
+import Language.Haskell.Extension
+         ( Extension(..), Language(..), knownLanguages, knownExtensions )
+import Text.EditDistance ( defaultEditCosts, levenshteinDistance )
+
+showUnsupportedExtension :: Extension -> String
+showUnsupportedExtension (EnableExtension knownExt) = prettyShow knownExt ++ "which is not supported"
+showUnsupportedExtension (DisableExtension knownExt) = prettyShow knownExt ++ "which is not supported"
```
Comment (phadej) [inline cabal-install-solver/src/Distribution/Solver/Modular/MessageUtils.hs:21]: I'm not sure whether

```
showUnsupportedExtension e@(DisableExtension _) = prettyShow e ++ ...
```

would be better to have `No` in the output. Fine either way.

Comment (skykanin) [reply cabal-install-solver/src/Distribution/Solver/Modular/MessageUtils.hs]: I think that would be better. Fixed

File: cabal-install-solver/src/Distribution/Solver/Modular/MessageUtils.hs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+import Data.Foldable (minimumBy)
+import Data.Ord (comparing)
+import Distribution.Pretty (prettyShow) -- from Cabal
+import Language.Haskell.Extension
+         ( Extension(..), Language(..), knownLanguages, knownExtensions )
+import Text.EditDistance ( defaultEditCosts, levenshteinDistance )
+
+showUnsupportedExtension :: Extension -> String
+showUnsupportedExtension (EnableExtension knownExt) = prettyShow knownExt ++ "which is not supported"
+showUnsupportedExtension extension@(DisableExtension _) = prettyShow extension ++ "which is not supported"
```
Comment (phadej) [inline cabal-install-solver/src/Distribution/Solver/Modular/MessageUtils.hs:21]: maybe format it in the same way as `showUnsupportedLanguage`?

Sorry for nitpicking on minor things.

Comment (skykanin) [reply cabal-install-solver/src/Distribution/Solver/Modular/MessageUtils.hs:21]: Ah yeah, that makes sense.

Comment (skykanin) [reply cabal-install-solver/src/Distribution/Solver/Modular/MessageUtils.hs]: Fixed.

**Comment by skykanin** [comment]
Since I've added a new test suit to `cabal-install-solver`

https://github.com/skykanin/cabal/blob/20af6ee736d5c2a74b16b9112fce18d0f8ced798/cabal-install-solver/cabal-install-solver.cabal#L128

Is it necessary to edit some file or cinfig to make the CI run these unit-tests as well?

**Review by emilypi — APPROVED** [review]
LGTM

**Comment by phadej** [comment]
> Since I've added a new test suit to `cabal-install-solver`
> 
> https://github.com/skykanin/cabal/blob/20af6ee736d5c2a74b16b9112fce18d0f8ced798/cabal-install-solver/cabal-install-solver.cabal#L128
> 
> Is it necessary to edit some file or config to make the CI run these new unit tests along with the others?

Great question.

Yes, `validate.sh`. If you have docker then using `make validate-via-docker-8.6.5` (for example) is one way to run it (clean docker-provided environment has less moving parts). Make your test-suite fail and try to make that command fail as well. IMO you don't need to do that though, you could open issue for maintainers to do that.

**Comment by Mikolaj** [comment]
Regarding

> 20:20 < skykanin> https://github.com/haskell/cabal/pull/7473 ready for merge I think

I earlier asked

> 10:06 < skykanin> I think my cabal PR can be merged now by someone who has write access https://github.com/haskell/cabal/pull/7473
> 10:20 < mikolaj__> skykanin: splendid; what's your decision regarding 'new test suite to cabal-install-solver'?
> 10:22 < mikolaj__> also, any preference which of the 3 merge buttons to use (since you have a merge commit among your commits, I'd default to rebase in the hopes it gets nuked out by rebase)

Posting here in case your IRC client doesn't preserve backlog.

**Comment by skykanin** [comment]
> Regarding
> 
> > 20:20 < skykanin> #7473 ready for merge I think
> 
> I earlier asked
> 
> > 10:06 < skykanin> I think my cabal PR can be merged now by someone who has write access #7473
> > 10:20 < mikolaj__> skykanin: splendid; what's your decision regarding 'new test suite to cabal-install-solver'?
> > 10:22 < mikolaj__> also, any preference which of the 3 merge buttons to use (since you have a merge commit among your commits, I'd default to rebase in the hopes it gets nuked out by rebase)
> 
> Posting here in case your IRC client doesn't preserve backlog.

I think I'll open a separate issue for the new test suit and I'd probably do rebase as well

**Comment by Mikolaj** [comment]
@skykanin: thank you and congratulations!

**Pull request merged**